### PR TITLE
O3-3426: Distro Inheritance Tool for the Build-Distro plugin

### PR DESCRIPTION
**Pull request by wikumChamith**
## Description of what I changed

Adding distro inheritance support to the build-distro plugin.

## Issue I worked on

see https://openmrs.atlassian.net/browse/O3-3431

## Checklist: I completed these to help reviewers :)

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean install` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute the above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/BuildDistro.java [hunk 1]
```diff
@@ -198,6 +202,18 @@ public void executeTask() throws MojoExecutionException, MojoFailureException {
 						+ buildDirectory.getAbsolutePath() + "\n");
 	}
 
+	private DistroProperties handleParentDistroProperties(Artifact distroArtifact, File buildDirectory, DistroProperties distroProperties) throws MojoExecutionException {
+		if (StringUtils.isNotBlank(distroArtifact.getArtifactId()) && StringUtils.isNotBlank(distroArtifact.getGroupId()) && StringUtils.isNotBlank(distroArtifact.getVersion())) {
+			Server tempServer = new Server.ServerBuilder().build();
```
Comment (ibacher) [inline maven-plugin/src/main/java/org/openmrs/maven/plugins/BuildDistro.java:207]: Is there a way we could de-couple this from the server setup? It seems like the only real reason we pass this in is for the server directory so maybe `resolveParentArtifact()` and `getArtifactProperties()` could be updated to just take a directory as an argument?

**Review by ibacher — COMMENTED** [review]
Is the code here recursive, i.e., if my parent distro has a parent is that also processed?

**Comment by wikumChamith** [comment]
> Is the code here recursive, i.e., if my parent distro has a parent is that also processed?

@ibacher, this implementation isn't recursive. Once it's in place, how can I test it by creating a parent distro that also has its parent?

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/BuildDistro.java (hunk 1 above)
Comment (wikumChamith) [reply maven-plugin/src/main/java/org/openmrs/maven/plugins/BuildDistro.java]: I updated the PR with these changes.

**Review by ibacher — APPROVED** [review]

**Comment by ibacher** [comment]
> Once it's in place, how can I test it by creating a parent distro that also has its parent?

Hmmm... I'm not entirely sure. I suppose if we have / added a way to resolve to a local directory rather than via Maven that might work? Alternatively, we can create dummy distro artifacts...

**Comment by wikumChamith** [comment]
> > Once it's in place, how can I test it by creating a parent distro that also has its parent?
> 
> Hmmm... I'm not entirely sure. I suppose if we have / added a way to resolve to a local directory rather than via Maven that might work? Alternatively, we can create dummy distro artifacts...

@ibacher can you provide more information on how to do that?

**Comment by ibacher** [comment]
I think it would be something like the bundled modules we already have for the "build" integration tests. Basically, we'd have some "bundled" dummy distributions parent1 and parent2 where parent2 depends on parent1. Then, before running a test for them, we use Maven to install those distributions in the local repo, and hopefully the tooling should just work.

**Comment by wikumChamith** [comment]
> I think it would be something like the bundled modules we already have for the "build" integration tests. Basically, we'd have some "bundled" dummy distributions parent1 and parent2 where parent2 depends on parent1. Then, before running a test for them, we use Maven to install those distributions in the local repo, and hopefully the tooling should just work.

Thank you. I'll merge this and move on to that in a separate PR.

**Pull request merged**